### ISSUE TITLE
"csc.exe" exited with code -2146232797 when constructing an invalid implicitly type array of Func

### ISSUE THREAD

**Issue by pasauter** [dotnet/roslyn#71593]
Visual Studio Version 17.8.2
C# Compiler version 4.8.0-7.23572.1

Compiler seems to crash instead of giving legit compiler error when constructing an implicitly typed array of Func with an invalid reference in one of the array elements.

Sample code:
```
namespace ClassLibrary1
{
    public class Class1
    {
        IEnumerable<Func<object>> test = new[]
        {
            () => CompileError,
        };
    }
}
```

Build Log:
[msbuild.log](https://github.com/dotnet/roslyn/files/13909630/msbuild.log)

**Comment by jcouv** [comment]
Tried in unittest shown below, in VS and in [sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAFgG4CYIwBTAZwAcIBjBkgYQBsImTADJRgAJwhiAnnnwBvAiSVkAzGUw9+g2cpIL8u3eRUAedAFYTAe2AArBuwAuAPmclHzRyQC8JGAwB3AG0AXUVDJX0IiIAKAEofN24rMBYoXgYAUTExKzE4cIiAXxoDZSKCIqA). But the problem did not repro for any of these environments. 
From the build log, it looks like this issue occurred for the user even with a recent version of the SDK (the csc.exe command shows references to .NET 8.0). So I'm not sure what's going on.

@pasauter Would you be able to capture a crash dump? That would help show where in the code the crash occurred. Here are some [instructions](https://github.com/dotnet/roslyn/blob/main/docs/wiki/Troubleshooting-tips.md#using-a-registry-setting).

```
    [Fact]
    public void TODO2()
    {
        string source = """
using System;
using System.Collections.Generic;

namespace ClassLibrary1
{
    public class Class1
    {
        IEnumerable<Func<object>> test = new[]
        {
            () => CompileError,
        };
    }
}
""";
        var comp = CreateCompilation(source, targetFramework: TargetFramework.Net80);
        comp.VerifyEmitDiagnostics(
            // (10,19): error CS0103: The name 'CompileError' does not exist in the current context
            //             () => CompileError,
            Diagnostic(ErrorCode.ERR_NameNotInContext, "CompileError").WithArguments("CompileError").WithLocation(10, 19)
            ); // no crash...
    }
```

**Comment by pasauter** [comment]
Sure thing: [dump file](https://www.dropbox.com/scl/fi/sxv2ol6yy8jx4s8iqdh73/csc.exe.36056.dmp?rlkey=iw52mxoshfui7cooxq3j69a66&dl=0)

**Comment by jcouv** [comment]
Thanks much for the dump file. Looking at the stacktrace for the exception (copied below), I see this is the same issue as https://github.com/dotnet/roslyn/issues/71002 which we fixed recently. The fix should ship in VS 17.9.
I'll go ahead and resolve this issue as duplicate. Thanks for reporting the problem

```
This program location is thought to be unreachable. File='/_/src/Compilers/CSharp/Portable/Symbols/FunctionTypeSymbol.cs' Line=146
   at Microsoft.CodeAnalysis.CSharp.Symbols.FunctionTypeSymbol.CreateITypeSymbol(NullableAnnotation nullableAnnotation)
   at Microsoft.CodeAnalysis.CSharp.BoundExpression.GetPublicTypeSymbol()
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundConversionOperation(BoundConversion boundConversion, Boolean forceOperandImplicitLiteral)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.Create(BoundNode boundNode)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateFromArray[TBoundNode,TOperation](ImmutableArray`1 boundNodes)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundArrayInitializationOperation(BoundArrayInitialization boundArrayInitialization)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.Create(BoundNode boundNode)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundArrayCreationOperation(BoundArrayCreation boundArrayCreation)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.Create(BoundNode boundNode)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundConversionOperation(BoundConversion boundConversion, Boolean forceOperandImplicitLiteral)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.Create(BoundNode boundNode)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundFieldEqualsValueOperation(BoundFieldEqualsValue boundFieldEqualsValue)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.Create(BoundNode boundNode)
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetRootOperation()
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetOperationWorker(CSharpSyntaxNode node, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetOperationWorker(CSharpSyntaxNode node, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetOperationCore(SyntaxNode node, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.GetOperationBlocksToAnalyze(ImmutableArray`1 executableBlocks, SemanticModel semanticModel, CancellationToken cancellationToken) in Microsoft.CodeAnalysis.Diagnostics\AnalyzerDriver.cs:line 2426
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.<>c__DisplayClass11_0.<ExecuteDeclaringReferenceActions>g__executeExecutableCodeActions|2() in Microsoft.CodeAnalysis.Diagnostics\AnalyzerDriver.cs:line 2246
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.ExecuteDeclaringReferenceActions(SyntaxReference decl, SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, GroupedAnalyzerActions coreActions, GroupedAnalyzerActions additionalPerSymbolActions, Boolean shouldExecuteSyntaxNodeActions, Boolean shouldExecuteOperationActions, Boolean shouldExecuteCodeBlockActions, Boolean shouldExecuteOperationBlockActions, Boolean isInGeneratedCode, CancellationToken cancellationToken) in Microsoft.CodeAnalysis.Diagnostics\AnalyzerDriver.cs:line 2343
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.ExecuteDeclaringReferenceActions(SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, Boolean isGeneratedCodeSymbol, IGroupedAnalyzerActions additionalPerSymbolActions, CancellationToken cancellationToken) in Microsoft.CodeAnalysis.Diagnostics\AnalyzerDriver.cs:line 2133
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.<TryProcessSymbolDeclaredAsync>d__137.MoveNext() in Microsoft.CodeAnalysis.Diagnostics\AnalyzerDriver.cs:line 1350
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.<TryProcessEventCoreAsync>d__136.MoveNext() in Microsoft.CodeAnalysis.Diagnostics\AnalyzerDriver.cs:line 1330
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.<ProcessEventAsync>d__134.MoveNext() in Microsoft.CodeAnalysis.Diagnostics\AnalyzerDriver.cs:line 1237
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.<ProcessCompilationEventsCoreAsync>d__133.MoveNext() in Microsoft.CodeAnalysis.Diagnostics\AnalyzerDriver.cs:line 1218
   ```
